### PR TITLE
cgroups.py: Add filters for the tasks() function

### PR DESCRIPTION
**Pull request by Elieva**
Closes #241 
The tasks() function allows to get the tasks that are in a cgroup.
Filters for the tasks TID, name and cmdline have been added to the
parameters of the function such that it is possible to select the tasks
that match these patterns.

Signed-off-by: Elieva Pignat <Elieva.Pignat@arm.com>

### CODE REVIEW

File: devlib/module/cgroups.py [hunk 1]
```diff
@@ -168,7 +169,31 @@ def move_all_tasks_to(self, dest, exclude=[]):
...
+        Report the tasks that are included in a cgroup. The tasks can be
+        filtered by their tid, tname or tcmdline if filter_tid, filter_tname or
+        filter_tcmdline are defined respectively. In this case, the reported
+        tasks are the ones in the cgroup that match these patterns.
+
+        Example of tasks format:
+        TID,tname,tcmdline
+        903,cameraserver,/system/bin/cameraserver
+
+        :params filter_tid: pattern to filter by TID
```
Comment (derkling) [inline devlib/module/cgroups.py]: Maybe we should specify that's a regexp?

File: devlib/module/cgroups.py [hunk 2]
```diff
@@ -180,13 +205,19 @@ def tasks(self, cgroup):
         tasks = {}
         for task in entries:
             tid = task.split(',')[0]
+            if filter_tid and not re.search(filter_tid, tid):
```
Comment (derkling) [inline devlib/module/cgroups.py]: Just to be a bit more sure it's a "sort of regexp, should we use instead:
```python
if type(filter_tid) != str and re.search(filter_tid, tid):
```

File: devlib/module/cgroups.py [hunk 3]
```diff
@@ -180,13 +205,19 @@ def tasks(self, cgroup):
...
             try:
                 tname = task.split(',')[1]
+                if filter_tname and not re.search(filter_tname, tname):
+                    continue
             except: continue
             try:
                 tcmdline = task.split(',')[2]
             except:
                 tcmdline = ''
+            if filter_tcmdline and not re.search(filter_tcmdline, tcmdline):
```
Comment (derkling) [inline devlib/module/cgroups.py]: This should be in the try block above... we don't want to filter the empty string.

File: devlib/module/cgroups.py (hunk 1 above)
Comment (derkling) [reply devlib/module/cgroups.py]: Should we also add a sanity check? It should be either None or a string...

File: devlib/module/cgroups.py (hunk 2 above)
Comment (derkling) [reply devlib/module/cgroups.py]: You can skip the check on type if you do the sanity check at the beginning...

File: devlib/module/cgroups.py (hunk 1 above)
Comment (Elieva) [reply devlib/module/cgroups.py]: specify that is a regex: Done
i have added sanity checks at the start of the function for the filter. I change as well the default value of the filter for the empty string.

File: devlib/module/cgroups.py (hunk 2 above)
Comment (Elieva) [reply devlib/module/cgroups.py]: Done: sanity check at the start of the function

File: devlib/module/cgroups.py (hunk 3 above)
Comment (Elieva) [reply devlib/module/cgroups.py]: I am not sure about this one. If someone wants to filter some pattern in the cmdline, it makes sense to me that a task that has no cmdline should not be in the result.

File: devlib/module/cgroups.py [hunk 4]
```diff
@@ -180,13 +211,19 @@ def tasks(self, cgroup):
         tasks = {}
         for task in entries:
```
Comment (setrofim) [inline devlib/module/cgroups.py]: I would re-write this loop to split only once for each task at the beginning: 

```
tid_str, tname, tcmdline = task.split(',', 2)
tid = int(tid_str)
```

(also note the limit on the split to handle the case where cmdline itself contains commas).

File: devlib/module/cgroups.py [hunk 5]
```diff
@@ -168,7 +169,37 @@ def move_all_tasks_to(self, dest, exclude=[]):
...
+        :type filter_tcmdline: str
+
+        :returns: a dictionary in the form: {tid:(tname, tcmdline)}
+        """
+        if type(filter_tid) != str:
+            raise TypeError('filter_tid should be a str')
+        if type(filter_tname) != str:
+            raise TypeError('filter_tname should be a str')
+        if type(filter_tcmdline) != str:
+            raise TypeError('filter_tcmdline should be a str')
```
[the remark this answers is not in the document]
Comment (Elieva) [reply devlib/module/cgroups.py]: I think it is better to keep filter_tid as a string such that a regexp can be given which allows more flexible filtering (not only equality)

Change type() by isinstance(): Done

File: devlib/module/cgroups.py (hunk 4 above)
Comment (Elieva) [reply devlib/module/cgroups.py]: Done

File: devlib/module/cgroups.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (setrofim) [reply devlib/module/cgroups.py]: >I think it is better to keep filter_tid as a string such that a regexp can be given which allows more flexible filtering (not only equality)

OK, makes sense.

**Pull request merged**